### ISSUE TITLE
Improve single-use daemon messaging

### ISSUE THREAD

**Issue by big-guy** [gradle/gradle#2660]
If `JAVA_OPTS` or `GRADLE_OPTS` do not match our defaults for the daemon or the build sets `org.gradle.jvmargs`, when someone runs with `--no-daemon`, we fork a single use Gradle daemon.

We print this message:

> To honour the JVM settings for this build a new JVM will be forked. Please consider using the daemon: https://docs.gradle.org/4.2-20170801184915+0000/userguide/gradle_daemon.html.
> Daemon will be stopped at the end of the build stopping after processing

There are a few confusing things here:
- "the JVM settings" -- which "JVM settings" are these?
- We say "Daemon will be stopped at the end of the build" and "consider using the daemon"

I think we should do these things:
1. Reword the "single use daemon" message.  Something like "A single use Gradle daemon will be used because the current JVM is incompatible. Learn more: <url>".  We should remove the technical "Daemon will be stopped at the end of the build stopping after processing" message.
2. The Gradle daemon chapter should be reviewed.  There is still some language that appears to assume the daemon is disabled by default. 
3. We should add a new "single user daemon" section to the Gradle daemon chapter. It should describe in more detail why a single use daemon is created.  We can also include information about how to avoid it.
4. We should fix the build-scan plugin so that it does not report the single use daemon as a good thing: https://github.com/gradle/gradle/issues/2597

**Comment by adammurdoch** [comment]
I would think about removing the single use daemon, and do one of the following:

1. Warn when the JVM args defined in your environment ($JAVA_OPTS/$GRADLE_OPTS/command-line/~/.gradle/gradle.properties/whatever) are not compatible with those specified in the build definition (in `gradle.properties`/whatever), but go ahead with using the definition from the environment. Also provide some diagnostics, such as where the environment definition came from and what it resolved to, where the build definition came from and what it resolved to, and in which ways they are incompatible.
2. Or, fail in the above case, and tell you how to proceed.
3. Or, warn but go ahead with using the definition from the build.
4. Or, silently prefer whatever is specified in the environment.
5. Or, silently prefer whatever is specified in the build definition.

**Comment by adammurdoch** [comment]
We should also consider making this consistent with what happens when you run with the daemon and the build definition and environment definition conflict with each other.

**Comment by adammurdoch** [comment]
actually, 'no single use daemon' and 'prefer the build definition' won't work together, scratch those options.

**Comment by Stono** [comment]
I just stumbled across this btw, we don't use the gradle daemon because we're building apps in docker containers, so single use is what we want.  Would be great if we could suppress the warning.

**Comment by nngo** [comment]
My `GRADLE_OPTS` has `-Dorg.gradle.daemon=false` to disable the daemon for a CI build, how do I disable the `To honour the JVM settings for this build a new JVM will be forked. Please consider using the daemon: ...` message?

**Comment by arnaud-deprez** [comment]
> I would think about removing the single use daemon, and do one of the following:
> 
>     1. Warn when the JVM args defined in your environment ($JAVA_OPTS/$GRADLE_OPTS/command-line/~/.gradle/gradle.properties/whatever) are not compatible with those specified in the build definition (in `gradle.properties`/whatever), but go ahead with using the definition from the environment. Also provide some diagnostics, such as where the environment definition came from and what it resolved to, where the build definition came from and what it resolved to, and in which ways they are incompatible.
> 
>     2. Or, fail in the above case, and tell you how to proceed.
> 
>     3. Or, warn but go ahead with using the definition from the build.
> 
>     4. Or, silently prefer whatever is specified in the environment.
> 
>     5. Or, silently prefer whatever is specified in the build definition.

I'm definitely for option 1.

If the user says 

> I don't want a gradle daemon by setting `org.gradle.daemon=false` or passing `--no-daemon`

Then gradle should not create a daemon or fork but notify him what settings it resolves in case of conflicts.

**Comment by sdeleuze** [comment]
Please allow us to remove this irritating warning. It is perfectly reasonable for users to disable Gradle daemon via `org.gradle.daemon=false` and prefer to wait a little bit longer when Gradle starts. No need to punish us ;-)

**Comment by queryholic** [comment]
please fix this bug quickly :)

**Comment by avonengel** [comment]
Could someone point out which settings need to match so that Gradle does not fork a single-use daemon?
Example:
I created a new project through IntelliJ's wizard with Kotlin DSL:

```gradle
plugins {
    kotlin("jvm") version "1.3.70"
}

group = "org.example"
version = "1.0-SNAPSHOT"

repositories {
    mavenCentral()
}

dependencies {
    implementation(kotlin("stdlib-jdk8"))
}

tasks {
    compileKotlin {
        kotlinOptions.jvmTarget = "1.8"
    }
    compileTestKotlin {
        kotlinOptions.jvmTarget = "1.8"
    }
}
```

When I run this inside an `openjdk/8-alpine` container, I get that message. I have no clue why that is, though...

**Comment by chanseokoh** [comment]
I cannot figure out the condition that is triggering a single-use daemon. I set neither `JAVA_OPTS`, `GRADLE_OPTS`, nor `org.gradle.jvmarg`. I don't have `~/.gradle/gradle.properties` or `gradle.properties` in my project. Nothing I pass on the command line except `--no-daemon`.

I do need to disable the daemon, but I have no clue as to how.

**Comment by ndw** [comment]
If I've explicitly disabled the daemon, could you please not print out that damned "Please consider using the daemon" message and all other messages related to it. I don't want to use it. On several occasions when I tried to use it, I got build errors or incorrect builds that worked just fine if I disabled it. I have neither the experience with gradle nor the interest necessary to help debug the problems, nor do I especially care about making the builds faster. I just want them to be correct. And I don't need to be asked to use it every time because the answer is always going to be "no".

**Comment by chanseokoh** [comment]
Regarding https://github.com/gradle/gradle/issues/2660#issuecomment-601854215, here's what worked as a workaround.

First, run with `--no-daemon --debug` and the check what JVM arguments are being used. They will be shown at the beginning of the output.
```
14:27:17.775 [INFO] [org.gradle.internal.nativeintegration.services.NativeServices] Initialized native services in: /home/chanseok/.gradle/native
14:27:17.957 [LIFECYCLE] [org.gradle.launcher.daemon.client.SingleUseDaemonClient] To honour the JVM settings for this build a new JVM will be forked. Please consider using the daemon: https://docs.gradle.org/5.2.1/userguide/gradle_daemon.html.
14:27:17.987 [DEBUG] [org.gradle.launcher.daemon.client.DefaultDaemonStarter] Using daemon args: [/home/chanseok/jdk1.8.0_171/bin/java, -XX:MaxMetaspaceSize=256m, -XX:+HeapDumpOnOutOfMemoryError, -Xmx512m, -Dfile.encoding=UTF-8, -Duser.country=US, -Duser.language=en, -Duser.variant, -cp, /home/chanseok/gradle-5.2.1/lib/gradle-launcher-5.2.1.jar]
```
In my case, they were
```
-XX:MaxMetaspaceSize=256m
-XX:+HeapDumpOnOutOfMemoryError
-Xmx512m
-Dfile.encoding=UTF-8
-Duser.country=US
-Duser.language=en
-Duser.variant
```
Then, use the exact same arguments via `GRADLE_OPTS` on the command line. That is,

```
GRADLE_OPTS="-XX:MaxMetaspaceSize=256m
             -XX:+HeapDumpOnOutOfMemoryError
             -Xmx512m
             -Dfile.encoding=UTF-8
             -Duser.country=US
             -Duser.language=en
             -Duser.variant" \
    ./gradlew --no-daemon --debug
```
Now it runs without using a daemon.
```
10:23:22.608 [INFO] [org.gradle.internal.nativeintegration.services.NativeServices] Initialized native services in: /home/chanseok/.gradle/native                                                                  
10:23:23.281 [INFO] [org.gradle.internal.work.DefaultWorkerLeaseService] Using 8 worker leases.                                                                                                                    
10:23:23.337 [DEBUG] [org.gradle.cache.internal.DefaultCacheAccess] Creating new cache for fileHashes, path /home/chanseok/.gradle/caches/5.1/fileHashes/fileHashes.bin, access org.gradle.cache.internal.DefaultCacheAccess@4aa83f4f                                                                                                                                                                                                 
```

**Comment by chanseokoh** [comment]
Actually, looks like I don't need to pass `-D` arguments. The following worked in my case.
```
GRADLE_OPTS="-XX:MaxMetaspaceSize=256m
             -XX:+HeapDumpOnOutOfMemoryError
             -Xmx512m" ./gradlew --no-daemon --debug
```
However, removing `-Xmx512m` causes Gradle to use a single-use daemon.

**Comment by leemeador** [comment]
The link in the error message about the "single use daemon" seems to point to a non-existent link for version 6.3. Is that expected behavior?

```
To honour the JVM settings for this build a new JVM will be forked. Please consider using the daemon: https://docs.gradle.org/6.3/userguide/gradle_daemon.html.
Daemon will be stopped at the end of the build stopping after processing
```

**Comment by garkin** [comment]
For the future pepegas who'll end here during google skirmishes against Gradle configuration riddles. (e.g. me 6 months later)

About debugging Graddle tasks.

```
gradle -Dorg.gradle.debug=true --no-daemon jar
```
- Property `-Dorg.gradle.debug=true` is not the same as `--debug` (which is just a tracing logging).
- With `--no-daemon` *it starts a daemon* and pauses building until the debugger will be attached through 5005 port. 
- Yes, `--no-daemon` is still needed or there won't be any pause, but you *should ignore warning* about daemon being forced by JVM args.
- Nothing happens in the console because **daemon waits for the remote debugger** to be attached. 

https://www.jetbrains.com/help/idea/work-with-gradle-tasks.html#debug_gradle
https://docs.gradle.org/7.0/userguide/build_environment.html#sec:gradle_configuration_properties
https://docs.gradle.org/7.0/userguide/gradle_daemon.html

---
Sort of on topic, additional console warning about this daemon state would be great.

Also `-Dorg.gradle.debug=true` docs doesn't say anything about `--no-daemon`. And it looks like it does nothing without it.

**Comment by ljacomet** [comment]
This is not in scope for the moment.

**Comment by MrFoxPro** [comment]
This is one of the reason I hate all about Java ecosystem. 
I want to fucking build my APK with 50 LOC single source file and 2 dependencies, instead I forced to learn what is android studio, how to deal with gradle, why it doesn't take into account provided arguments.
I really hope someday we will able to throw out all this unusable 'enterprise' piece of shit that requires to learn multiple chapters of abstractions and special language that will never be usable outside of ecosystem and which takes a lot RAM and CPU usage to do basic things.

**Comment by garkin** [comment]
@MrFoxPro, dear sir. I'm not sure about ecosystem you came from, but a better person, which you may or may not want to become sometimes in the future, would try to respect other participants. Please, watch your language and keep your frustration at bay.

**Comment by s4nji** [comment]
Going insane from this, I cannot get it to run without the daemon.

For anyone still looking into this still, this is where the check whether single-use-daemon should be used happens:  
https://github.com/gradle/gradle/blob/01a5dfa7c63a8055672eae1aa235f603810a8e4a/platforms/core-runtime/gradle-cli/src/main/java/org/gradle/launcher/cli/BuildActionsFactory.java#L148

And this is the full comparison list:  
https://github.com/gradle/gradle/blob/cccbb046ef9c2f791298a42c7998c90e789c79aa/platforms/core-runtime/daemon-protocol/src/main/java/org/gradle/launcher/daemon/context/DaemonCompatibilitySpec.java#L42-L55

As you can see, the java opts is just one of the several checks done.  
It even returns a nice string why it doesnt satisfy the conditions, but it is never printed.

**Comment by ljacomet** [comment]
We still intend to remove the no-daemon support from Gradle and only have single use or long live daemons in the future.

Part of this is improved by having a better explanation in why Gradle spawned a daemon, see the following issue released in Gradle 9.5.0:
* #33212

**Comment by rocketraman** [comment]
> We still intend to remove the no-daemon support from Gradle and only have single use or long live daemons in the future.

@ljacomet  here are quite a few use cases people have identified that this change would affect. Have all of these been considered in this decision?

See for example:

* https://github.com/gradle/gradle/issues/16110#issuecomment-2140689842
* https://github.com/gradle/gradle/issues/16110#issuecomment-2168204703
* https://github.com/gradle/gradle/issues/16110#issuecomment-2256406142
* https://github.com/gradle/gradle/issues/16110#issuecomment-2710740325
